pkg/cmd/experimental/syncgroups/ad/ldapinterface.go

```diff
@@ -0,0 +1,147 @@
+
+	// userQuery holds the information necessary to make an LDAP query for a specific
+	// first-class user entry on the LDAP server
+	userQuery ldaputil.LDAPQueryOnAttribute
+	// groupMembershipAttributes defines which attributes on an LDAP user entry will be interpreted as its ldapGroupUID
+	groupMembershipAttributes []string
+	// UserNameAttributes defines which attributes on an LDAP user entry will be interpreted as its name
+	userNameAttributes []string
+
+	cachePopulated bool
+	// cachedUsers holds the result of user queries for later reference, indexed on user UID
+	// e.g. this will map an LDAP user UID to the LDAP entry returned from the query made using it
+	cachedUsers map[string]*ldap.Entry
+
+	ldapGroupToLDAPMembers map[string][]*ldap.Entry
+}
+
+var _ ldapinterfaces.LDAPMemberExtractor = &LDAPInterface{}
+var _ ldapinterfaces.LDAPGroupLister = &LDAPInterface{}
+
+// ExtractMembers returns the LDAP member entries for a group specified with a ldapGroupUID
+func (e *LDAPInterface) ExtractMembers(ldapGroupUID string) ([]*ldap.Entry, error) {
+	if !e.cachePopulated {
+		// This happens in cases where we did not list out every group.  In that case, we're going to be asked about specific groups.
+		// don't worry about caching, there ought not be duplicate in this list
+
+		usersInGroup := []*ldap.Entry{}
+
+		// check for all users with ldapGroupUID in any of the allowed member attributes
+		for _, currAttribute := range e.groupMembershipAttributes {
+			currQuery := e.userQuery
+			currQuery.QueryAttribute = currAttribute
+
+			searchRequest, err := currQuery.NewSearchRequest(ldapGroupUID, e.requiredUserAttributes())
+			if err != nil {
+				return nil, err
+			}
+
+			currEntries, err := ldaputil.QueryForEntries(e.clientConfig, searchRequest)
+			if err != nil {
+				return nil, err
+			}
+
+			for i := range currEntries {
+				currEntry := currEntries[i]
+				for j := range usersInGroup {
+					existingUser := usersInGroup[j]
+					if existingUser.DN == currEntry.DN {
+						continue
+					}
+					usersInGroup = append(usersInGroup, existingUser)
+				}
+			}
+		}
+
+		return usersInGroup, nil
+	}
+
+	return e.ldapGroupToLDAPMembers[ldapGroupUID], nil
+}
+
+// ListGroups queries for all groups as configured with the common group filter and returns their
+// LDAP group UIDs. This also satisfies the LDAPGroupLister interface
+func (e *LDAPInterface) ListGroups() ([]string, error) {
+	if err := e.populateCache(); err != nil {
+		return nil, err
+	}
+
+	return sets.KeySet(reflect.ValueOf(e.ldapGroupToLDAPMembers)).List(), nil
+}
+
+// populateCache queries all users to build a map of all the groups.  If the cache has already been
+// populated, this is a no-op.
+func (e *LDAPInterface) populateCache() error {
+	if e.cachePopulated {
+		return nil
+	}
+
+	searchRequest := e.userQuery.LDAPQuery.NewSearchRequest(e.requiredUserAttributes())
+
+	userEntries, err := ldaputil.QueryForEntries(e.clientConfig, searchRequest)
```

Comment:
if userEntries is an empty slice do we still want to consider the cache populated?